File: app/src/main/java/com/silas/omaster/data/util/WatermarkRenderer.kt
package com.silas.omaster.data.util

import android.content.Context
import android.graphics.*
import android.media.MediaScannerConnection
import android.net.Uri
import com.silas.omaster.data.model.*
import com.silas.omaster.ui.theme.HasselbladColors
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * 水印渲染器
 * 负责将水印模板渲染到图片上
 */
class WatermarkRenderer(private val context: Context) {

    private val exifProvider = ExifWatermarkProvider(context)
    private val dateFormat = SimpleDateFormat("yyyy.MM.dd", Locale.getDefault())

    /**
     * 导出带水印的图片
     */
    suspend fun exportWatermarkedImage(
        sourceUri: Uri,
        template: MasterWatermarkTemplate,
        format: ExportFormat = ExportFormat.JPEG,
        quality: Int = 95
    ): File = withContext(Dispatchers.IO) {
        // 1. 加载原图
        val sourceBitmap = loadBitmap(sourceUri)
        
        // 2. 提取 EXIF
        val exifData = exifProvider.extractFromUri(sourceUri)
        
        // 3. 应用水印
        val watermarked = render(sourceBitmap, template, exifData)
        
        // 4. 写入文件
        val outputDir = File(context.cacheDir, "watermarked")
        outputDir.mkdirs()
        val outputFile = File(outputDir, "OMaster_${System.currentTimeMillis()}.${format.extension}")
        
        outputFile.outputStream().use { stream ->
            watermarked.compress(format.compressFormat, quality, stream)
        }
        
        // 5. 通知相册刷新
        MediaScannerConnection.scanFile(context, arrayOf(outputFile.absolutePath), null, null)
        
        // 6. 回收 Bitmap
        if (watermarked != sourceBitmap) {
            watermarked.recycle()
        }
        sourceBitmap.recycle()
        
        outputFile
    }

    /**
     * 渲染水印
     */
    fun render(
        sourceBitmap: Bitmap,
        template: MasterWatermarkTemplate,
        exifData: ExifWatermarkData
    ): Bitmap {
        // 创建可变 Bitmap
        val result = sourceBitmap.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(result)
        
        // 根据模板 ID 选择渲染方式
        when (template.id) {
            "hasselblad-master" -> renderHasselbladMaster(canvas, result, exifData)
            "hasselblad-hncs" -> renderHasselbladHncs(canvas, result, exifData)
            "hasselblad-xpan" -> renderHasselbladXpan(canvas, result, exifData)
            else -> renderGeneric(canvas, result, template, exifData)
        }
        
        return result
    }

    /**
     * 渲染哈苏大师印记
     */
    private fun renderHasselbladMaster(
        canvas: Canvas,
        bitmap: Bitmap,
        exifData: ExifWatermarkData
    ) {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()
        
        // 底部信息栏高度 (8% 画面高度)
        val barHeight = height * 0.08f
        val barTop = height - barHeight
        
        // 绘制半透明背景
        val bgPaint = Paint().apply {
            color = HasselbladColors.BackgroundSemiTransparent.toArgb()
            style = Paint.Style.FILL
        }
        canvas.drawRect(0f, barTop, width, height, bgPaint)
        
        // 文字画笔
        val textPaint = Paint().apply {
            isAntiAlias = true
            textAlign = Paint.Align.CENTER
        }
        
        // 金色分割线
        val dividerPaint = Paint().apply {
            color = HasselbladColors.Gold.toArgb()
            strokeWidth = 1f
            alpha = 77 // 30%
        }
        val dividerY = barTop + barHeight * 0.15f
        canvas.drawLine(width * 0.2f, dividerY, width * 0.8f, dividerY, dividerPaint)
        
        // HASSELBLAD 品牌
        textPaint.apply {
            textSize = barHeight * 0.25f
            color = HasselbladColors.Gold.toArgb()
            typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
            letterSpacing = 0.1f
        }
        val brandY = barTop + barHeight * 0.4f
        canvas.drawText("HASSELBLAD", width / 2, brandY, textPaint)
        
        // HNCS 认证标识
        textPaint.apply {
            textSize = barHeight * 0.12f
            color = HasselbladColors.TextSecondary.toArgb()
            letterSpacing = 0.05f
        }
        canvas.drawText("HNCS", width / 2, brandY + barHeight * 0.2f, textPaint)
        
        // 设备型号
        textPaint.apply {
            textSize = barHeight * 0.15f
            color = HasselbladColors.TextSecondary.toArgb()
            typeface = Typeface.DEFAULT
            letterSpacing = 0f
        }
        val deviceInfo = exifData.getDeviceInfo()
        canvas.drawText(deviceInfo, width / 2, brandY + barHeight * 0.4f, textPaint)
        
        // 拍摄参数
        textPaint.apply {
            textSize = barHeight * 0.12f
            typeface = Typeface.MONOSPACE
            color = HasselbladColors.TextTertiary.toArgb()
        }
        val paramsInfo = exifData.getParamsInfo()
        canvas.drawText(paramsInfo, width / 2, brandY + barHeight * 0.6f, textPaint)
        
        // 日期和位置
        val dateStr = exifData.dateTaken ?: dateFormat.format(Date())
        textPaint.apply {
            textSize = barHeight * 0.1f
            typeface = Typeface.DEFAULT
        }
        canvas.drawText(dateStr, width / 2, barTop + barHeight * 0.85f, textPaint)
        
        // 底部品牌联合标识
        textPaint.apply {
            textSize = barHeight * 0.08f
            color = HasselbladColors.TextTertiary.toArgb()
        }
        canvas.drawText("OPPO × Hasselblad | Master Edition", width / 2, barTop + barHeight * 0.95f, textPaint)
    }

    /**
     * 渲染 HNCS 认证标识
     */
    private fun renderHasselbladHncs(
        canvas: Canvas,
        bitmap: Bitmap,
        exifData: ExifWatermarkData
    ) {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()
        
        // 徽章位置 (左上角)
        val badgeWidth = width * 0.15f
        val badgeHeight = height * 0.08f
        val badgeLeft = width * 0.05f
        val badgeTop = height * 0.05f
        
        // 绘制徽章背景
        val bgPaint = Paint().apply {
            color = HasselbladColors.BackgroundSemiTransparent.toArgb()
            style = Paint.Style.FILL
        }
        val cornerRadius = badgeWidth * 0.1f
        val rectF = RectF(badgeLeft, badgeTop, badgeLeft + badgeWidth, badgeTop + badgeHeight)
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, bgPaint)
        
        // 绘制金色边框
        val borderPaint = Paint().apply {
            color = HasselbladColors.Gold.toArgb()
            style = Paint.Style.STROKE
            strokeWidth = 2f
        }
        canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, borderPaint)
        
        // 绘制文字
        val textPaint = Paint().apply {
            isAntiAlias = true
            textAlign = Paint.Align.CENTER
            color = HasselbladColors.Gold.toArgb()
            textSize = badgeHeight * 0.3f
            typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        }
        
        // 👑 HNCS
        canvas.drawText("HNCS", badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.4f, textPaint)
        
        // 哈苏自然色彩认证
        textPaint.apply {
            textSize = badgeHeight * 0.2f
            color = HasselbladColors.TextSecondary.toArgb()
            typeface = Typeface.DEFAULT
        }
        canvas.drawText("哈苏自然", badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.65f, textPaint)
        canvas.drawText("色彩认证", badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.85f, textPaint)
    }

    /**
     * 渲染 XPAN 宽幅印记
     */
    private fun renderHasselbladXpan(
        canvas: Canvas,
        bitmap: Bitmap,
        exifData: ExifWatermarkData
    ) {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()
        
        // 底部信息条高度
        val barHeight = height * 0.05f
        val barTop = height - barHeight
        
        // 绘制背景
        val bgPaint = Paint().apply {
            color = Color.BLACK
            alpha = 180
            style = Paint.Style.FILL
        }
        canvas.drawRect(0f, barTop, width, height, bgPaint)
        
        // 文字画笔
        val textPaint = Paint().apply {
            isAntiAlias = true
            textAlign = Paint.Align.LEFT
        }
        
        // XPAN 65:24
        textPaint.apply {
            textSize = barHeight * 0.5f
            color = HasselbladColors.TextPrimary.toArgb()
            typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
            letterSpacing = 0.15f
        }
        val leftMargin = width * 0.05f
        canvas.drawText("XPAN  65:24", leftMargin, barTop + barHeight * 0.7f, textPaint)
        
        // HASSELBLAD · 设备型号
        textPaint.apply {
            textAlign = Paint.Align.RIGHT
            textSize = barHeight * 0.35f
            color = HasselbladColors.TextSecondary.toArgb()
            typeface = Typeface.DEFAULT
            letterSpacing = 0f
        }
        val rightMargin = width * 0.95f
        val deviceInfo = "HASSELBLAD · ${exifData.getDeviceInfo()}"
        canvas.drawText(deviceInfo, rightMargin, barTop + barHeight * 0.65f, textPaint)
    }

    /**
     * 渲染通用水印
     */
    private fun renderGeneric(
        canvas: Canvas,
        bitmap: Bitmap,
        template: MasterWatermarkTemplate,
        exifData: ExifWatermarkData
    ) {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()
        
        // 获取模板图层
        val layers = template.layers.sortedBy { it.sortOrder }
        
        // 计算位置
        val position = template.defaultPosition
        val (x, y, align) = getPositionCoords(position, width, height)
        
        // 文字画笔
        val textPaint = Paint().apply {
            isAntiAlias = true
            textAlign = align
        }
        
        // 绘制每个图层
        var currentY = y
        layers.forEach { layer ->
            if (!layer.isEnabled) return@forEach
            
            val content = getLayerContent(layer, exifData)
            if (content.isEmpty()) return@forEach
            
            textPaint.apply {
                textSize = layer.defaultStyle.fontSize * (height / 1000f)
                color = Color.parseColor(layer.defaultStyle.color)
                alpha = (layer.defaultStyle.opacity * 255).toInt()
                letterSpacing = layer.defaultStyle.letterSpacing / 100f
                if (layer.defaultStyle.fontWeight >= 700) {
                    typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                } else {
                    typeface = Typeface.DEFAULT
                }
            }
            
            canvas.drawText(content, x, currentY, textPaint)
            currentY += textPaint.textSize * 1.5f
        }
    }

    /**
     * 获取图层内容
     */
    private fun getLayerContent(layer: WatermarkLayerDef, exifData: ExifWatermarkData): String {
        return when (layer.contentSource) {
            ContentSource.MANUAL -> layer.defaultContent
            ContentSource.EXIF -> when (layer.type) {
                WatermarkLayerType.PARAMS -> exifData.getParamsInfo()
                WatermarkLayerType.TIMESTAMP -> exifData.dateTaken ?: ""
                else -> layer.defaultContent
            }
            ContentSource.DEVICE_INFO -> exifData.getDeviceInfo()
            ContentSource.SYSTEM -> dateFormat.format(Date())
            ContentSource.GPS -> exifData.locationName ?: ""
        }
    }

    /**
     * 获取位置坐标
     */
    private fun getPositionCoords(
        position: WatermarkPosition,
        width: Float,
        height: Float
    ): Triple<Float, Float, Paint.Align> {
        val padding = width * 0.05f
        return when (position) {
            WatermarkPosition.TOP_LEFT -> Triple(padding, padding * 2, Paint.Align.LEFT)
            WatermarkPosition.TOP_CENTER -> Triple(width / 2, padding * 2, Paint.Align.CENTER)
            WatermarkPosition.TOP_RIGHT -> Triple(width - padding, padding * 2, Paint.Align.RIGHT)
            WatermarkPosition.CENTER_LEFT -> Triple(padding, height / 2, Paint.Align.LEFT)
            WatermarkPosition.CENTER -> Triple(width / 2, height / 2, Paint.Align.CENTER)
            WatermarkPosition.CENTER_RIGHT -> Triple(width - padding, height / 2, Paint.Align.RIGHT)
            WatermarkPosition.BOTTOM_LEFT -> Triple(padding, height - padding, Paint.Align.LEFT)
            WatermarkPosition.BOTTOM_CENTER -> Triple(width / 2, height - padding, Paint.Align.CENTER)
            WatermarkPosition.BOTTOM_RIGHT -> Triple(width - padding, height - padding, Paint.Align.RIGHT)
        }
    }

    /**
     * 加载 Bitmap
     */
    private fun loadBitmap(uri: Uri): Bitmap {
        val inputStream = context.contentResolver.openInputStream(uri)
            ?: throw IllegalArgumentException("Cannot open URI: $uri")
        return BitmapFactory.decodeStream(inputStream).also {
            inputStream.close()
        }
    }
}

/**
 * 导出格式
 */
enum class ExportFormat(val extension: String, val compressFormat: Bitmap.CompressFormat) {
    JPEG("jpg", Bitmap.CompressFormat.JPEG),
    PNG("png", Bitmap.CompressFormat.PNG),
    WEBP("webp", Bitmap.CompressFormat.WEBP)
}

File: src/services/watermarkRenderer.ts
/**
 * 水印渲染器
 * Web端 - 使用Canvas渲染水印并导出图片
 */

import { MasterWatermarkTemplate, ExifWatermarkData, WatermarkPosition } from '../types/watermark';
import { HasselbladColors, calculateImageLuminance, getTextColorForBackground } from './hasselbladColors';

// ========== 导出格式 ==========
export type ExportFormat = 'jpeg' | 'png' | 'webp';

export interface ExportOptions {
  format: ExportFormat;
  quality: number;  // 0-100
  filename?: string;
}

// ========== 水印渲染器类 ==========
export class WatermarkRenderer {
  private canvas: HTMLCanvasElement;
  private ctx: CanvasRenderingContext2D;

  constructor() {
    this.canvas = document.createElement('canvas');
    const ctx = this.canvas.getContext('2d');
    if (!ctx) throw new Error('Cannot get canvas context');
    this.ctx = ctx;
  }

  /**
   * 导出带水印的图片
   */
  async exportWatermarkedImage(
    imageSource: string | HTMLImageElement,
    template: MasterWatermarkTemplate,
    exifData: Partial<ExifWatermarkData>,
    options: ExportOptions = { format: 'jpeg', quality: 95 }
  ): Promise<Blob> {
    // 1. 加载图片
    const img = await this.loadImage(imageSource);
    
    // 2. 设置画布尺寸
    this.canvas.width = img.width;
    this.canvas.height = img.height;
    
    // 3. 绘制原图
    this.ctx.drawImage(img, 0, 0);
    
    // 4. 渲染水印
    this.render(template, exifData);
    
    // 5. 导出为 Blob
    return new Promise((resolve, reject) => {
      this.canvas.toBlob(
        (blob) => {
          if (blob) resolve(blob);
          else reject(new Error('Failed to create blob'));
        },
        `image/${options.format}`,
        options.quality / 100
      );
    });
  }

  /**
   * 导出并下载
   */
  async exportAndDownload(
    imageSource: string | HTMLImageElement,
    template: MasterWatermarkTemplate,
    exifData: Partial<ExifWatermarkData>,
    options: ExportOptions = { format: 'jpeg', quality: 95 }
  ): Promise<void> {
    const blob = await this.exportWatermarkedImage(imageSource, template, exifData, options);
    
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = options.filename || `OMaster_${Date.now()}.${options.format}`;
    a.click();
    URL.revokeObjectURL(url);
  }

  /**
   * 渲染水印
   */
  render(
    template: MasterWatermarkTemplate,
    exifData: Partial<ExifWatermarkData>
  ): void {
    const { width, height } = this.canvas;

    // 根据模板 ID 选择渲染方式
    switch (template.id) {
      case 'hasselblad-master':
        this.renderHasselbladMaster(exifData);
        break;
      case 'hasselblad-hncs':
        this.renderHasselbladHncs(exifData);
        break;
      case 'hasselblad-xpan':
        this.renderHasselbladXpan(exifData);
        break;
      default:
        this.renderGeneric(template, exifData);
    }
  }

  /**
   * 渲染哈苏大师印记
   */
  private renderHasselbladMaster(exifData: Partial<ExifWatermarkData>): void {
    const { width, height } = this.canvas;
    const ctx = this.ctx;

    // 底部信息栏高度 (8% 画面高度)
    const barHeight = height * 0.08;
    const barTop = height - barHeight;

    // 绘制半透明背景
    ctx.fillStyle = HasselbladColors.BackgroundSemiTransparent;
    ctx.fillRect(0, barTop, width, barHeight);

    // 金色分割线
    ctx.strokeStyle = HasselbladColors.Gold;
    ctx.globalAlpha = 0.3;
    ctx.lineWidth = 1;
    const dividerY = barTop + barHeight * 0.15;
    ctx.beginPath();
    ctx.moveTo(width * 0.2, dividerY);
    ctx.lineTo(width * 0.8, dividerY);
    ctx.stroke();
    ctx.globalAlpha = 1;

    // HASSELBLAD 品牌
    ctx.fillStyle = HasselbladColors.Gold;
    ctx.font = `bold ${barHeight * 0.25}px sans-serif`;
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.letterSpacing = '2px';
    const brandY = barTop + barHeight * 0.4;
    ctx.fillText('HASSELBLAD', width / 2, brandY);

    // HNCS 认证标识
    ctx.fillStyle = HasselbladColors.TextSecondary;
    ctx.font = `${barHeight * 0.12}px sans-serif`;
    ctx.fillText('HNCS', width / 2, brandY + barHeight * 0.2);

    // 设备型号
    const deviceInfo = this.getDeviceInfo(exifData);
    ctx.fillStyle = HasselbladColors.TextSecondary;
    ctx.font = `${barHeight * 0.15}px sans-serif`;
    ctx.fillText(deviceInfo, width / 2, brandY + barHeight * 0.4);

    // 拍摄参数
    const paramsInfo = this.getParamsInfo(exifData);
    ctx.fillStyle = HasselbladColors.TextTertiary;
    ctx.font = `${barHeight * 0.12}px monospace`;
    ctx.fillText(paramsInfo, width / 2, brandY + barHeight * 0.6);

    // 日期
    const dateStr = exifData.dateTaken || this.getCurrentDate();
    ctx.fillStyle = HasselbladColors.TextTertiary;
    ctx.font = `${barHeight * 0.1}px sans-serif`;
    ctx.fillText(dateStr, width / 2, barTop + barHeight * 0.85);

    // 底部品牌联合标识
    ctx.fillStyle = HasselbladColors.TextTertiary;
    ctx.font = `${barHeight * 0.08}px sans-serif`;
    ctx.fillText('OPPO × Hasselblad | Master Edition', width / 2, barTop + barHeight * 0.95);
  }

  /**
   * 渲染 HNCS 认证标识
   */
  private renderHasselbladHncs(exifData: Partial<ExifWatermarkData>): void {
    const { width, height } = this.canvas;
    const ctx = this.ctx;

    // 徽章位置 (左上角)
    const badgeWidth = width * 0.15;
    const badgeHeight = height * 0.08;
    const badgeLeft = width * 0.05;
    const badgeTop = height * 0.05;
    const cornerRadius = badgeWidth * 0.1;

    // 绘制徽章背景
    ctx.fillStyle = HasselbladColors.BackgroundSemiTransparent;
    this.roundRect(badgeLeft, badgeTop, badgeWidth, badgeHeight, cornerRadius);
    ctx.fill();

    // 绘制金色边框
    ctx.strokeStyle = HasselbladColors.Gold;
    ctx.lineWidth = 2;
    this.roundRect(badgeLeft, badgeTop, badgeWidth, badgeHeight, cornerRadius);
    ctx.stroke();

    // HNCS 文字
    ctx.fillStyle = HasselbladColors.Gold;
    ctx.font = `bold ${badgeHeight * 0.3}px sans-serif`;
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillText('HNCS', badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.35);

    // 哈苏自然色彩认证
    ctx.fillStyle = HasselbladColors.TextSecondary;
    ctx.font = `${badgeHeight * 0.2}px sans-serif`;
    ctx.fillText('哈苏自然', badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.6);
    ctx.fillText('色彩认证', badgeLeft + badgeWidth / 2, badgeTop + badgeHeight * 0.8);
  }

  /**
   * 渲染 XPAN 宽幅印记
   */
  private renderHasselbladXpan(exifData: Partial<ExifWatermarkData>): void {
    const { width, height } = this.canvas;
    const ctx = this.ctx;

    // 底部信息条高度
    const barHeight = height * 0.05;
    const barTop = height - barHeight;

    // 绘制背景
    ctx.fillStyle = 'rgba(0, 0, 0, 0.7)';
    ctx.fillRect(0, barTop, width, barHeight);

    // XPAN 65:24
    ctx.fillStyle = HasselbladColors.TextPrimary;
    ctx.font = `bold ${barHeight * 0.5}px sans-serif`;
    ctx.textAlign = 'left';
    ctx.textBaseline = 'middle';
    ctx.letterSpacing = '3px';
    const leftMargin = width * 0.05;
    ctx.fillText('XPAN  65:24', leftMargin, barTop + barHeight * 0.5);

    // HASSELBLAD · 设备型号
    const deviceInfo = `HASSELBLAD · ${this.getDeviceInfo(exifData)}`;
    ctx.fillStyle = HasselbladColors.TextSecondary;
    ctx.font = `${barHeight * 0.35}px sans-serif`;
    ctx.textAlign = 'right';
    ctx.letterSpacing = '0px';
    const rightMargin = width * 0.95;
    ctx.fillText(deviceInfo, rightMargin, barTop + barHeight * 0.5);
  }

  /**
   * 渲染通用水印
   */
  private renderGeneric(
    template: MasterWatermarkTemplate,
    exifData: Partial<ExifWatermarkData>
  ): void {
    const { width, height } = this.canvas;
    const ctx = this.ctx;

    const layers = template.layers.sort((a, b) => a.sortOrder - b.sortOrder);
    const position = template.defaultPosition;
    const { x, y, align } = this.getPositionCoords(position, width, height);

    ctx.textAlign = align;
    ctx.textBaseline = 'top';

    let currentY = y;
    for (const layer of layers) {
      if (!layer.isEnabled) continue;

      const content = this.getLayerContent(layer, exifData);
      if (!content) continue;

      const fontSize = layer.defaultStyle.fontSize * (height / 1000);
      ctx.font = `${layer.defaultStyle.fontWeight >= 700 ? 'bold' : 'normal'} ${fontSize}px ${layer.defaultStyle.fontFamily}`;
      ctx.fillStyle = layer.defaultStyle.color;
      ctx.globalAlpha = layer.defaultStyle.opacity;

      ctx.fillText(content, x, currentY);
      currentY += fontSize * 1.5;
    }

    ctx.globalAlpha = 1;
  }

  // ========== 辅助方法 ==========

  private async loadImage(source: string | HTMLImageElement): Promise<HTMLImageElement> {
    if (source instanceof HTMLImageElement) {
      return source;
    }

    return new Promise((resolve, reject) => {
      const img = new Image();
      img.crossOrigin = 'anonymous';
      img.onload = () => resolve(img);
      img.onerror = () => reject(new Error('Failed to load image'));
      img.src = source;
    });
  }

  private roundRect(
    x: number,
    y: number,
    w: number,
    h: number,
    r: number
  ): void {
    const ctx = this.ctx;
    ctx.beginPath();
    ctx.moveTo(x + r, y);
    ctx.lineTo(x + w - r, y);
    ctx.quadraticCurveTo(x + w, y, x + w, y + r);
    ctx.lineTo(x + w, y + h - r);
    ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
    ctx.lineTo(x + r, y + h);
    ctx.quadraticCurveTo(x, y + h, x, y + h - r);
    ctx.lineTo(x, y + r);
    ctx.quadraticCurveTo(x, y, x + r, y);
    ctx.closePath();
  }

  private getDeviceInfo(exif: Partial<ExifWatermarkData>): string {
    if (exif.make && exif.model) return `${exif.make} ${exif.model}`;
    if (exif.model) return exif.model;
    return 'Unknown Device';
  }

  private getParamsInfo(exif: Partial<ExifWatermarkData>): string {
    const parts: string[] = [];
    if (exif.aperture) parts.push(exif.aperture);
    if (exif.shutterSpeed) parts.push(exif.shutterSpeed);
    if (exif.iso) parts.push(exif.iso);
    if (exif.focalLength) parts.push(exif.focalLength);
    return parts.join('  ');
  }

  private getCurrentDate(): string {
    const now = new Date();
    return `${now.getFullYear()}.${String(now.getMonth() + 1).padStart(2, '0')}.${String(now.getDate()).padStart(2, '0')}`;
  }

  private getLayerContent(
    layer: { type: string; defaultContent: string; contentSource: string },
    exif: Partial<ExifWatermarkData>
  ): string {
    switch (layer.contentSource) {
      case 'manual':
        return layer.defaultContent;
      case 'exif':
        if (layer.type === 'params') return this.getParamsInfo(exif);
        if (layer.type === 'timestamp') return exif.dateTaken || '';
        return layer.defaultContent;
      case 'device_info':
        return this.getDeviceInfo(exif);
      case 'system':
        return this.getCurrentDate();
      case 'gps':
        return exif.locationName || '';
      default:
        return layer.defaultContent;
    }
  }

  private getPositionCoords(
    position: WatermarkPosition,
    width: number,
    height: number
  ): { x: number; y: number; align: CanvasTextAlign } {
    const padding = width * 0.05;
    switch (position) {
      case 'top-left':
        return { x: padding, y: padding * 2, align: 'left' };
      case 'top-center':
        return { x: width / 2, y: padding * 2, align: 'center' };
      case 'top-right':
        return { x: width - padding, y: padding * 2, align: 'right' };
      case 'center-left':
        return { x: padding, y: height / 2, align: 'left' };
      case 'center':
        return { x: width / 2, y: height / 2, align: 'center' };
      case 'center-right':
        return { x: width - padding, y: height / 2, align: 'right' };
      case 'bottom-left':
        return { x: padding, y: height - padding, align: 'left' };
      case 'bottom-center':
        return { x: width / 2, y: height - padding, align: 'center' };
      case 'bottom-right':
        return { x: width - padding, y: height - padding, align: 'right' };
    }
  }
}

// ========== 单例导出 ==========
export const watermarkRenderer = new WatermarkRenderer();

// ========== 便捷方法 ==========
export async function exportWatermarkedImage(
  imageSource: string | HTMLImageElement,
  template: MasterWatermarkTemplate,
  exifData: Partial<ExifWatermarkData>,
  options?: ExportOptions
): Promise<Blob> {
  return watermarkRenderer.exportWatermarkedImage(imageSource, template, exifData, options);
}

export async function exportAndDownload(
  imageSource: string | HTMLImageElement,
  template: MasterWatermarkTemplate,
  exifData: Partial<ExifWatermarkData>,
  options?: ExportOptions
): Promise<void> {
  return watermarkRenderer.exportAndDownload(imageSource, template, exifData, options);
}
